src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,860 @@
+                } else {
+                    // this shouldn't be possible - because all outputs from connected subgraphs will have a tmpVar declared.
+                }
+            }
+        }
+
+        if (callString.endsWith(', ')) callString = callString.slice(0, -2);
+
+        callString += ' );\n';
+
+        return callString;
+    },
+
+    addStaticSwitch: function (name, id, maxValue, value) {
+        if (this._switchMap[name]) {
+            console.error('switch ' + name + ' already added!');
+        } else {
+            this._switchMap[name] = { id: id, maxValue: maxValue, value: value };
+            this._switchId2Name[id] = name;
+        }
+    },
+
+    getSwitchValue: function (name) {
+        if (this._switchMap[name]) {
+            return ((this._switchOverrides && this._switchOverrides[name] !== undefined) ? this._switchOverrides[name] : this._switchMap[name].value);
+        }
+    },
+
+    setSwitchValue: function (name, value) {
+        if (this._switchMap[name]) {
+            if (value !== undefined && value >= 0 && value <= this._switchMap[name].maxValue) {
+                if (this._switchMap[name].value != value) {
+                    this._switchMap[name].value = value;
+                    this._defaultParamsDirty = true;
+                    this._defaultSwitchesDirty = true;
+                }
+            }
+        }
+    },
+
+    getSwitchKey: function () {
+        var switchNames = Object.keys(this._switchMap);
+
+        var key = "graph_" + this.id;
+
+        for (var k = 0; k < switchNames.length; k++) {
+            var switchName = switchNames[k];
+            key += "switch_" + this._switchMap[switchName].id;
+            key += ":" + this.getSwitchValue(switchName);
+        }
+
+        return key;
+    },
+
+    update: function (device) {
+        if (this._defaultParamsDirty === true) {
+            this.updateUniforms();
+            this.setParameters(device);
+            this._defaultParamsDirty = false;
+        }
+        if (this._defaultSwitchesDirty === true) {
+            // any material using this graph directly will need to switch shader variant
+            for (var m = 0; m < this._materials.length; m++) {
+                this._materials[m].dirty = true;
+                this._materials[m].dirtyShader = true;
+            }
+            this._defaultSwitchesDirty = false;
+        }
+    },
+
+    setParameter: function (name, value) {
+        var ioPort = this.getIoPortByName(name);
+        if (ioPort && value !== undefined) {
+            if (!((value instanceof Texture || typeof(value) === 'number') ? value === ioPort._value : value.equals(ioPort._value))) {
+                this._addIoPort(ioPort.type, name, value, ioPort.ptype );
+            }
+        }
+    },
+
+    getIoPortByName: function (name) {
+        return (this._portMap[name] === undefined) ? undefined : this.graphData.ioPorts[this._portMap[name]];

```

Comment:
this._portMap[name] should not execute twice

Reply:
I will add a var

Reply:
improved